flocker/control/test/test_model_change.py

```diff
@@ -0,0 +1,418 @@
+# Copyright ClusterHQ Inc.  See LICENSE file for details.
+
+"""
+Whenever the configuration model changes we need to write code to
+upgrade the on-disk format from previous releases. This module will
+automatically detect such changes by failing a test, ensuring that upgrade
+code is always implemented when necessary.
+"""
+
+from json import loads, dumps
+from subprocess import check_output
+
+from pyrsistent import (
+    PRecord, PClass, CheckedPSet, CheckedPVector, CheckedPMap, field,
+)
+
+from twisted.trial.unittest import SynchronousTestCase
+from twisted.python.filepath import FilePath
+from twisted.python.reflect import qual as fqpn
+
+from .._model import Deployment
+
+PERSISTED_MODEL = FilePath(__file__).sibling(b"persisted_model.json")
+
+# The class at the root of the configuration tree. This may need to be
+# changed if the configuration's root class changes.
+ROOT_CLASS = Deployment
+
+
+def generate_model(root_class=ROOT_CLASS):
+    """
+    Generate a data-structure that represents the current configuration
+    model.
+
+    Changes to output may require regenerating the persisted version
+    on-disk.
+
+    This should switch to Pyrsistent's introspection API once it exists:
+    https://github.com/tobgu/pyrsistent/issues/47
+    """
+    classes_result = {}
+    result = {u"root": fqpn(root_class),
+              u"classes": classes_result}
+    classes = {root_class}
+    while classes:
+        klass = classes.pop()
+        klass_name = fqpn(klass)
+        if klass_name in classes_result:
+            continue
+        record = None
+        if issubclass(klass, (PRecord, PClass)):
+            if issubclass(klass, PRecord):
+                attr_name = "_precord_fields"
+            else:
+                attr_name = "_pclass_fields"
+            record = {u"category": u"record",
+                      u"fields": {}}
+            for name, field_info in getattr(klass, attr_name).items():
+                record[u"fields"][name] = list(
+                    fqpn(cls) for cls in field_info.type)
```

Comment:
It seems that we can't rely on the order of `field_info.type` being stable.  The test suite is failing now, at least in part because:

```
              'flocker.control._model.Application': {u'category': u'record',
                                                     u'fields': {'command_line': ['flocker.control._model.UnicodePVector',
                                                                                  '__builtin__.NoneType'],
...
```
does not equal
```
              u'flocker.control._model.Application': {u'category': u'record',
                                                      u'fields': {u'command_line': [u'__builtin__.NoneType',
                                                                                    u'flocker.control._model.UnicodePVector'],
...
```

Maybe there are other places where order is non-deterministic.  I stopped looking for differences after I found one.

(Awful, awful, awful presentation of the failure, btw :/ Reminds me that we haven't yet upgraded to the version of Twisted that includes the better formatting functionality, though I don't know how much it would help in this specific case.)